Add Linux SCHED_DEADLINE support

RafaelOrtizRC opened the pull request:
This adds rudimentary support for [`SCHED_DEADLINE`](https://www.kernel.org/doc/Documentation/scheduler/sched-deadline.txt) on Linux systems. It is tested and working on `Linux 5.4.0-84-generic #94-Ubuntu SMP Thu Aug 26 20:27:37 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`, but I see no reason it wouldn't work as far back as `3.14` when `SCHED_DEADLINE` was introduced.

There's a few ugly bits:
* Since `SCHED_DEADLINE` isn't POSIX, there's a bunch of `#[cfg(target_os = "linux")]` all over the place, a dedicated `linux.rs` might be better. I also noticed some `libc::` dependencies that aren't present on macOS, so I don't know that `unix.rs` is fully POSIX anyway.
* The `ThreadId` passed in is interpreted as a `pid_t` instead of being translated properly (solutions to which are also hacky), since the syscall for deadline scheduling expects a kernel tid (userspace pid).

vityafx commented:
Thanks much for the contribution! Is there a way we can test this policy by any chance?

RafaelOrtizRC commented:
> Thanks much for the contribution! Is there a way we can test this policy by any chance?

Sure! I added a test for the deadline policy as-implemented (and cleaned up some formatting stuff - CLion doesn't default to `rustfmt`).

RafaelOrtizRC commented:
It looks like the `4.4` kernel on xenial (which is the GitHub Actions runner you're using) will refuse to admit any threads with deadline scheduling when runtime == deadline == period (even though that's supposed to be supported). I'm changing the interface to add a `ThreadPriority::Deadline(u64, u64, u64)` so the tests can specify values that the scheduler should consistently admit. Will probably push it up tomorrow.

RafaelOrtizRC commented:
Sorry, I got pulled into other work. Anyway, the tests should be passing now. I added the explicit deadline interface as described and tested (working) on Xenial.

RafaelOrtizRC commented:
Okay, _now_ it should be working. I removed the clippy warning in that one test (I think it's more clear with the identity multiplication in this specific case) and fixed the CI issues (hopefully).

vityafx commented on .github/workflows/ci.yml:
```diff
@@ -56,9 +56,13 @@ jobs:
+
+      - name: Test all features (other)
+        if: matrix.features == '' && runner.os != 'macOS'
+        run: sudo -E $HOME/bin/cargo test --all-features
```
Out of curiosity, why was this necessary?

RafaelOrtizRC replied:
macOS doesn't implement `SCHED_DEADLINE` since it's not part of POSIX. The test will only pass on Linux. It looks like the rest of the tests are failing due to cargo being installed in a different location than I expected, and `sudo` not being part of the environment. The test on Linux needs to be executed as root.

RafaelOrtizRC replied:
I'm not familiar enough with GitHub Actions to know what the correct fix is here. I assumed that the run command would be executed as a regular shell command inside a VM, but it appears that was wrong.

vityafx replied:
Can you try changing it to `sudo -E /usr/share/rust/.cargo/bin/cargo test --all-features` please?

RafaelOrtizRC replied:
Try again now.

RafaelOrtizRC replied:
Looks like that (mostly) worked, I pushed up a typo fix. Sorry for the spam, I'm not super familiar with GitHub Actions.

vityafx commented on src/unix.rs:
```diff
@@ -219,20 +281,54 @@ pub fn thread_schedule_policy() -> Result<ThreadSchedulePolicy, Error> {
+                    libc::SYS_sched_setattr,
+                    tid,
+                    &sched_attr as *const _,
+                    // we are not setting SCHED_FLAG_RECLAIM nor SCHED_FLAG_DL_OVERRUN
```
Just curious, why don't we set it? Is it possible to set it?

vityafx replied:
I am not using such a policy anywhere, do you think it is worth adding these options to the interface so that others could configure them?

vityafx replied:
Sure! Thanks.

vityafx commented:
Enjoy https://crates.io/crates/thread-priority/0.3.0 and thanks much for the contribution!
